### PR TITLE
Fix broken specs

### PR DESCRIPTION
**Pull request by marshall-lee**
Puma migration in f905d35 didn't go well. Some specs are failing due to the following problems:

- `GzipServlet` does not see `Zlib`.
- `GzipServlet` should respond with `Content-Length` (WEBRick did it automatically).
- `BodyReadback` should use `CONTENT_TYPE` instead of `HTTP_CONTENT_TYPE`.

The reason that we don't see them failing is that part of the specs were even **not running**: some of the spec examples sets `trap('INT')` that does the global exit (added in ba2161c).
One should never set `trap` handlers inside tests since they're global to the running process.

Such a trap handler overwrites the one set by RSpec (btw RSpec's `SIGINT` handler is already overwritten by Puma that's not good too).

If you run `rspec ./spec/session_spec.rb` you'd see:

```
Finished in 14.51 seconds (files took 0.46612 seconds to load)
23 examples, 0 failures
```

But that file contains more than 23 examples, it should be 88.

### CODE REVIEW

**Comment by marshall-lee** [comment]
@julik @toland 

It seems that HTTPS doesn't work with `fork()` not only with SecureTransport but with OpenSSL too.
I'm not sure it should really work with fork at all — sharing SSL state between processes always create problems.

also: https://stackoverflow.com/questions/15466809/libcurl-ssl-error-after-fork

Also I don't know how to fix specs that use `/wrongcontentlength` endpoint with HTTPS connection and I'm not sure they're really fixable.

**Comment by julik** [comment]
Ohlala

**Comment by marshall-lee** [comment]
I commented out three failing ssl specs which I doubt they're even repairable in fea7101.
Also I make `PatronTestServer` to run in a separate process instead of thread because Puma and RSpec handles `SIGINT` differently: `257e6d1`.

**Comment by marshall-lee** [comment]
@toland @julik WDYT?

**Comment by toland** [comment]
It looks fine to me. However, at this point, @julik knows this code far better than I do. I will defer to him.

**Comment by julik** [comment]
I was out sick. This is some great work @marshall-lee , thank you. I didn't realise threading was so problematic in this configuration. I can't agree that removing Webrick tests was a bad idea though because it was exquisitely hard to test "slow" responses with it anyway.

File: spec/session_spec.rb [hunk 1]
```diff
@@ -16,9 +16,9 @@
     forbidden_protos = %w( smb tftp imap smtp telnet dict ftp sftp scp file gopher )
     forbidden_protos.each do |forbidden_proto|
       it "should deny a #{forbidden_proto.upcase} request" do
-        @session.base_url = nil
         expect {
-          @session.get('%s://localhost' % forbidden_proto)
+          @session.base_url = '%s://localhost' % forbidden_proto
+          @session.get('')
```
Comment (julik) [inline spec/session_spec.rb]: Very minor, but why is this change necessary? Do these non-HTTP URLs refuse to work when passed as an argument to `#get` as-is?

File: spec/session_ssl_spec.rb [hunk 2]
```diff
@@ -214,14 +213,14 @@
...
-  it "should ignore a wrong Content-Length when asked to" do
+  xit "should ignore a wrong Content-Length when asked to" do
     expect {
       @session.ignore_content_length = true
       @session.get("/wrongcontentlength")
     }.to_not raise_error
   end
 
-  it "should fail by default with a Content-Length too high" do
+  xit "should fail by default with a Content-Length too high" do
```
Comment (julik) [inline spec/session_ssl_spec.rb]: This test fails because we do not set the correct `Content-Length` in the response in the test server? Wouldn't it be better to find out why it fails?

File: spec/support/test_server.rb [hunk 3]
```diff
@@ -3,18 +3,24 @@
...
 class PatronTestServer
-  APP = Rack::Builder.new { eval(File.read(File.dirname(__FILE__) + '/config.ru')) }
+  APP = Rack::Builder.new { eval(File.read(File.join(__dir__, 'config.ru'))) }
 
   def self.start(ssl = false, port = 9001 )
+    # Reset the RSpec's SIGINT handler that does not really terminate after
+    # the first Ctrl+C pressed.
+    # Useful to terminate the forked process before Puma is actually started:
+    # it happens when running one particular example that does not need Puma
+    # so the specs are in face finished before the Puma started.
```
Comment (julik) [inline spec/support/test_server.rb]: "in fact"?

Comment (marshall-lee) [reply spec/support/test_server.rb]: yep, thanks

File: spec/session_ssl_spec.rb (hunk 2 above)
Comment (marshall-lee) [reply spec/session_ssl_spec.rb]: I managed to make this test pass either by setting protocol version to 1.0 or by passing `Connection: close` header:
```ruby
@session.http_version = 'HTTPv1_0'
```
Or:
```
@session.get("/wrongcontentlength", { 'Connection' => 'close' })
```
Otherwise, Puma will hold keep-alive connection and eventually request will fail due to timeout. I did the same fix for non-ssl specs too. 

However, it works only for libcurl compiled against OpenSSL. At Travis CI, libcurl is compiled against GnuTLS and request fails with such an error:
```
<Patron::Error: GnuTLS recv error (-9): A TLS packet with unexpected length was received.>
```
I guess it's reasonable. Sending wrong content-length is a protocol violation and GunTLS just appears to be more strict than OpenSSL.

Comment (julik) [reply spec/session_ssl_spec.rb]: Gotcha. Then we should remove those tests IMO as they test divergent behavior for which kno "golden master" outcome can be established.

File: spec/session_spec.rb (hunk 1 above)
Comment (marshall-lee) [reply spec/session_spec.rb]: If `base_url` is already set, it'll be joined with `get` argument.

Comment (julik) [reply spec/session_spec.rb]: Then I would qualify this as a bug - if the `base_url` specifies, say, `http://something` but we perform `telnet://somewhereelse` explicitly then I'd say what gets supplied in `get` should take priority and override `base_url` for the duration of the call?

Comment (marshall-lee) [reply spec/session_spec.rb]: Yep, I agree on this

File: spec/session_ssl_spec.rb (hunk 2 above)
Comment (marshall-lee) [reply spec/session_ssl_spec.rb]: OK

File: spec/session_spec.rb (hunk 1 above)
Comment (marshall-lee) [reply spec/session_spec.rb]: But let's do it in a separate PR

**Comment by marshall-lee** [comment]
@julik I did all the final fixes

**Comment by julik** [comment]
Thank you for your changes!
